xpdan/callbacks.py

```diff
@@ -7,3 +17,145 @@ def start(self, doc):
+        rpath = resource['resource_path']
+        root = resource.get('root', '')
+        root = self.root_map.get(root, root)
+        if root:
+            rpath = os.path.join(root, rpath)
+        ret = handler(rpath, **kwargs)
+        self.handlers[key] = ret
+
+    def datum(self, doc):
+        self.datums[doc['datum_id']] = doc
+
+    def retrieve_datum(self, datum_id):
+        doc = self.datums[datum_id]
+        resource = self.resources[doc['resource']]
+        handler_class = self.handler_reg[resource['spec']]
+        key = (str(resource['uid']), handler_class.__name__)
+        return self.handlers[key](**doc['datum_kwargs'])
+
+    def fill_event(self, event, fields=True, inplace=True):
+        if fields is True:
+            fields = set(event['data'])
+        elif fields is False:
+            # if no fields, we got nothing to do!
+            # just return back as-is
+            return event
+        elif fields:
+            fields = set(fields)
+
+        if not inplace:
+            ev = _sanitize(event)
+            ev = copy.deepcopy(ev)
+        else:
+            ev = event
+        data = ev['data']
+        filled = ev['filled']
+        for k in set(data) & fields:
+            # Try to fill the data
+            try:
+                print('hiw')
+                print(k, data[k])
+                v = self.retrieve_datum(data[k])
+                print(k, data[k], v)
+                data[k] = v
+                filled[k] = True
+            # If retrieve fails keep going
+            except (ValueError, KeyError) as e:
+                raise e
+        return event
+
+    def event(self, doc):
+        ev = self.fill_event(doc, inplace=False)
+        return ev
+
+
+class ExportCallback(Retrieve):
+    """Callback to copy data to a new root"""
+
+    def __init__(self, new_root, handler_reg, root_map=None):
+        # TODO: fix this, since it is a dirty hack which only works for ADTIFF
+        super().__init__(handler_reg, root_map)
+        self.new_root = new_root
+        self.old_root = None
+
+    def resource(self, doc):
+        doc = copy.deepcopy(doc)
+        super().resource(doc)
+        self.old_root = doc['root']
+        doc.update(root=self.new_root)
+        return doc
+
+    def datum(self, doc):
+        super().datum(doc)
+        # retrieve the datum using path only handler?
+        resource = self.resources[doc['resource']]
+        handler_class = self.handler_reg[resource['spec']]
+        key = (str(resource['uid']), handler_class.__name__)
+        fin = self.handlers[key].get_file_list([doc['datum_kwargs']])[0]
+
+        # replace the root with the new root
+        fout = os.path.join(self.new_root, os.path.relpath(fin, self.old_root))
+
+        ensure_path_exists(os.path.dirname(fout))
+        # cp fin, new_path
+        shutil.copy2(fin, fout)
+
+    def event(self, doc):
+        # don't fill
+        return doc
+
+
+class RemoteExportCallback(ExportCallback):
```

Comment:
We could roll these two classes into one if we used `rsync` as the default file transfer protocol.